content/en/blog/_posts/2021-12-20-kubernetes-in-kubernetes-and-pxe-bootable-server-farm.md

```diff
@@ -0,0 +1,201 @@
+
+* [**generic/values.yaml**](https://github.com/kvaps/kubefarm/blob/v0.13.1/examples/generic/values.yaml)
+
+Here you can specify the parameters which are passed into the upstream Kubernetes-in-Kubernetes chart. In order for you control-plane to be accessible from the outside, it is enough to specify the IP address here, but if you wish, you can specify some DNS name here.
+
+In the PXE server configuration you can specify a timezone. You can also add an SSH key for logging in without a password (but you can also specify a password), as well as kernel modules and parameters that should be applied during booting the system.
+
+Next comes the **nodePools** configuration, i.e. the nodes themselves. If you've ever used a terraform module for gke, then this logic will remind you of it. Here you statically describe all nodes with a set of parameters:
+
+-   **Name** (hostname);
+    
+-   **MAC-addresses** — we have nodes with two network cards, and each one can boot from any of the MAC addresses specified here.
+    
+-   **IP-address**, which the DHCP server should issue to this node.
+    
+
+In this example, you have two pools: the first has five nodes, the second has only one, the second pool has also two tags assigned. Tags are the way to describe configuration for specific nodes. For example, you can add specific DHCP options for some pools, options for the PXE server for booting (e.g. here is debug option enabled) and set of **kubernetesLabels** and **kubernetesTaints** options. What does that mean?
+
+For example, in this configuration you have a second nodePool with one node. The pool has **debug** and **foo** tags assigned. Now see the options for **foo** tag in **kubernetesLabels**. This means that the m1c43 node will boot with these two labels and taint assigned. Everything seems to be simple. Now [let's try](https://asciinema.org/a/40728) this in practice.
+
+### Demo
+
+Go to [examples](https://github.com/kvaps/kubefarm/tree/v0.13.1/examples) and update previously deployed chart to Kubefarm. Just use the [generic](https://github.com/kvaps/kubefarm/tree/v0.13.1/examples/generic) parameters and look at the pods. You can see that a PXE server and one more job were added. This job essentially goes to the deployed Kubernetes cluster and creates a new token. Now it will run repeatedly every 12 hours to generate a new token, so that the nodes can connect to your cluster.
+
+[![](/images/blog/2021-12-20-kubernetes-in-kubernetes-and-pxe-bootable-server-farm/screenshot05.svg)](https://asciinema.org/a/407282?autoplay=1)
+
+In a [graphical representation](https://kvaps.github.io/images/posts/kubernetes-in-kubernetes-and-pxe-bootable-server-farm/Argo_CD_Applications_kubefarm-network.html), it looks about the same, but now apiserver started to be exposed outside.
```

Comment:
Link to https://kvaps.github.io/images/posts/kubernetes-in-kubernetes-and-pxe-bootable-server-farm/Argo_CD_Applications_kubefarm-network.html is a 404